src/lib.rs

```diff
@@ -55,10 +68,30 @@ pub struct Storage<T>(StorageRef<T>, PendingRef);
 /// The pointer also holds the storage alive and knows the index of the element to look up.
 pub struct Pointer<T> {
     index: usize,
+    epoch: Epoch,
     target: StorageRef<T>,
     pending: PendingRef,
 }
 
+impl<T> Pointer<T> {
+    /// Creates a new `WeakPointer` to this component.
+    pub fn downgrade(&self) -> WeakPointer<T> {
+        WeakPointer {
+            index: self.index,
+            epoch: self.epoch,
+            target: Arc::downgrade(&self.target),
+            pending: Arc::downgrade(&self.pending),
+        }
+    }
+}
+
+/// Weak variant of `Pointer`.
+pub struct WeakPointer<T> {
+    index: usize,
+    epoch: Epoch,
+    target: WeakStorageRef<T>,
+    pending: WeakPendingRef,
```

Comment:
I don't think we need to fiddle with the weak pending reference, let's just use `PendingRef` here